### PR TITLE
Make recvLen' in MuxBearer.Socket strict

### PR DESCRIPTION
**Pull request by coot**
There's a loop which collects all the bytestring in a lazy list.  This
patch forces the lists as well as received bytestrings which it holds.

This patch also forces the received MuxSDU.

### CODE REVIEW

File: network-mux/src/Network/Mux/Bearer/Socket.hs [hunk 1]
```diff
@@ -73,10 +74,10 @@ socketAsMuxBearer tracer sd = do
                   return (header {Mx.msBlob = blob}, ts)
 
       recvLen' :: Bool -> Int64 -> [BL.ByteString] -> IO BL.ByteString
-      recvLen' _ 0 bufs = return (BL.concat $ reverse bufs)
-      recvLen' waitingOnNxtHeader l bufs = do
+      recvLen' _ 0 !bufs = return (BL.concat $ reverse bufs)
+      recvLen' waitingOnNxtHeader l !bufs = do
```
Comment (dcoutts) [inline network-mux/src/Network/Mux/Bearer/Socket.hs]: Why does the bufs list need the `!` here? Every call for it already passes a value that's in WHNF.

File: network-mux/src/Network/Mux/Bearer/Socket.hs [hunk 2]
```diff
@@ -73,10 +74,10 @@ socketAsMuxBearer tracer sd = do
...
           traceWith tracer $ Mx.MuxTraceRecvStart $ fromIntegral l
-          buf <- Socket.recv sd l
+          !buf <- Socket.recv sd l
```
Comment (dcoutts) [inline network-mux/src/Network/Mux/Bearer/Socket.hs]: Wouldn't it be better to ensure `Socket.recv` return a result that is in normal form?

File: network-mux/src/Network/Mux/Bearer/Socket.hs [hunk 3]
```diff
@@ -61,8 +62,8 @@ socketAsMuxBearer tracer sd = do
           --say "read"
           --hexDump hbuf ""
           case Mx.decodeMuxSDU hbuf of
-              Left  e      -> throwM e
-              Right header -> do
+              Left  e       -> throwM e
+              Right !header -> do
```
Comment (dcoutts) [inline network-mux/src/Network/Mux/Bearer/Socket.hs]: Wouldn't it be better to ensure that `Mx.decodeMuxSDU` returns a result that is in normal form?

File: network-mux/src/Network/Mux/Bearer/Socket.hs (hunk 1 above)
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/Socket.hs]: Sure they are in WHNF, but this avoids wrapping them in thunks:

```
get :: Int -> IO [String]
get = go []
  where
    go xs 0 = pure $ xs
    go xs n = do
      x <- getLine
      go (x : xs) (pred n)
```

```
> x <- get 3
a
b
c
> seq x ()
> :sprint x ()
x = _ : _
```

But
```
get :: Int -> IO [String]
get = go []
  where
    go !xs 0 = pure $ xs
    go !xs n = do
      !x <- getLine
      go (x : xs) (pred n)
```

```
> x <- get 3
a
b
c
> seq x ()
> :sprint x
x = [('a' : _),('b' : _),('c' : _)]
```

File: network-mux/src/Network/Mux/Bearer/Socket.hs (hunk 2 above)
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/Socket.hs]: I think we only can have an implication `WHNF => NF` for values returned by a lambda. This bang will ensure that it is in WHNF. A test for `WHNF => NF` would be useful.

**Comment by coot** [comment]
Closing as there was no real difference.

**Pull request closed without merging**